### ISSUE TITLE
test performance

### ISSUE THREAD

**Issue by wuliaososhunhun** [scaldi/scaldi-play#18]
Hi, 

Is there any concerns about the test performance when we use scaladi-play in test? currently I just refactored the project and I found that the test time is significantly longer than before(double the time roughtly). In the tests, I use running(new TestServer)/running(new FakeApplication) a lot which means that the system will launch play server a lot. Seems like it takes large trunks of time to initialise the system while some expensive initialisation may happen such as inmemory database/ mock rabbitmq etc.

any ideas or solutions may help?

Thanks,
Yanyang

====
Here is the data:
[info] Passed: Total 337, Failed 0, Errors 0, Passed 337
[success] Total time: 231 s

**Comment by OlegIlyenko** [comment]
Thanks for sharing this! I definitely interested to find the root cause if the performance issues. Have you also used scaldi before the refactoring? I assume you are talking about play 2.3 -> 2.4 migration, is it correct?

**Comment by wuliaososhunhun** [comment]
Nope. Basically I mean the version before I use scaladi with the version used scaladi-play 0.5.4.
Both play 2.3.

Here is the new test:

the old version without scaladi:
Passed: Total 343, Failed 0, Errors 0, Passed 343
[success] Total time: 102 s, completed 11-May-2015 10:37:45

the old version with scaladi
[info] Passed: Total 337, Failed 0, Errors 0, Passed 337
[success] Total time: 244 s, completed 11-May-2015 11:16:20

the new version with scaladi (loads of code changes introduced)
[info] Passed: Total 343, Failed 0, Errors 0, Passed 343
[success] Total time: 127 s, completed 11-May-2015 10:53:27

The second test takes long time may be caused by some settings as i did not realise any performance related updates in code. Potentially may be caused by the changes of logging level as I have seen a lot of logs about server startup but no log level changes discovered so far. I will dig this if have time.



Thanks,
Yanyang

**Comment by OlegIlyenko** [comment]
Thanks for the feedback and sorry for such a delayed response (recently I was a little bit busy and also currently working on play 2.4 support). If you are using scaldi in a test (especially if you are including most of the application modules) then I would consider these tests to be more of an integration test which by definition will take more time than unit tests for example, since you are generally starting the whole application. One thing you can do is to reuse initialized injector between different tests. In this case you need to make sure that you don't reuse already initialised injectors (and pointed out in the docs http://scaldi.org/learn/#testing)

What kind of changes have you introduced in the new version of your application? The test execution of it looks comparable to the version without scaldi. Would be really interested to hear about your discoveries - maybe you can find some specific places where you can observe performance issues

**Comment by elmalto** [comment]
My 2.4 tests run slower by a factor of 3:

2.4:
Run completed in 5 minutes, 36 seconds.

2.3: 
Run completed in 1 minute, 45 seconds.

I feel like it might have to do with the controller injection because I noticed services getting started from controllers that I never used in the test. Switching to lazy val in controller solved the issue. Is there a way to inject my controllers lazily because although I might need an app running, I usually don't need all controllers.
